controllers/storageconsumer/storageconsumer_controller.go

```diff
@@ -52,14 +72,366 @@ func (r *StorageConsumerReconciler) Reconcile(ctx context.Context, request recon
 
 	r.Log.Info("Reconciling StorageConsumer.", "StorageConsumer", klog.KRef(request.Namespace, request.Name))
 
-	// TODO: Implement complete functionality of the controller
+	// Initalize the reconciler properties from the request
+	r.initReconciler(request)
+
+	if err := r.get(r.storageConsumer); err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("No StorageConsumer resource.", "StorageConsumer", klog.KRef(r.storageConsumer.Namespace, r.storageConsumer.Name))
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		r.Log.Error(err, "Failed to retrieve StorageConsumer.", "StorageConsumer", klog.KRef(r.storageConsumer.Namespace, r.storageConsumer.Name))
+		return reconcile.Result{}, err
+	}
+
+	// Reconcile changes to the cluster
+	result, reconcileError := r.reconcilePhases()
+
+	// Apply status changes to the StorageConsumer
+	statusError := r.Client.Status().Update(r.ctx, r.storageConsumer)
+	if statusError != nil {
+		r.Log.Info("Could not update StorageConsumer status.", "StorageConsumer", klog.KRef(r.storageConsumer.Namespace, r.storageConsumer.Name))
+	}
+
+	// Reconcile errors have higher priority than status update errors
+	if reconcileError != nil {
+		return result, reconcileError
+	} else if statusError != nil {
+		return result, statusError
+	} else {
+		return result, nil
+	}
+}
+
+func (r *StorageConsumerReconciler) initReconciler(request reconcile.Request) {
+	r.ctx = context.Background()
+	r.namespace = request.Namespace
+
+	r.storageConsumer = &ocsv1alpha1.StorageConsumer{}
+	r.storageConsumer.Name = request.Name
+	r.storageConsumer.Namespace = r.namespace
+
+	r.cephBlockPool = &rookCephv1.CephBlockPool{}
+	r.cephBlockPool.Name = fmt.Sprintf("%s-%s", "cephblockpool", r.storageConsumer.Name)
+	r.cephBlockPool.Namespace = r.namespace
+
+	r.cephClientRBDProvisioner = &rookCephv1.CephClient{}
+	r.cephClientRBDProvisioner.Name = fmt.Sprintf("%s-%s", "cephclient-rbd-provisioner", r.storageConsumer.Name)
+	r.cephClientRBDProvisioner.Namespace = r.namespace
+
+	r.cephClientRBDNode = &rookCephv1.CephClient{}
+	r.cephClientRBDNode.Name = fmt.Sprintf("%s-%s", "cephclient-rbd-node", r.storageConsumer.Name)
+	r.cephClientRBDNode.Namespace = r.namespace
+
+	r.cephClientHealthChecker = &rookCephv1.CephClient{}
+	r.cephClientHealthChecker.Name = fmt.Sprintf("%s-%s", "cephclient-health-checker", r.storageConsumer.Name)
+	r.cephClientHealthChecker.Namespace = r.namespace
+
+}
+
+func (r *StorageConsumerReconciler) reconcilePhases() (reconcile.Result, error) {
+
+	r.storageConsumer.Status.State = v1alpha1.StorageConsumerStateConfiguring
+
+	r.storageConsumer.Status.CephResources = nil
```

Comment:
I think this is erroneous, resetting the array means that if this reconcile fails at an early stage will update the StorageConsumer status with an empty status for the reconciled objects, but these objects might exist (and have status) from a prev reconcile invocation. 

I think we should keep prev status for existing resources that were not touched during the current reconciliation. 
To do so, you need to remove this line

Reply:
This has been removed and introduced logic for modifying the existing cephResources list.